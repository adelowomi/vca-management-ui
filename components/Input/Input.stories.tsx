import { action } from '@storybook/addon-actions';
import {
  boolean,
  number,
  select,
  text,
  withKnobs,
} from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import React from 'react';

import Input from './Input';

const options = {
  button: 'button',
  checkbox: 'checkbox',
  color: 'color',
  email: 'email',
  file: 'file',
  hidden: 'hidden',
  image: 'image',
  number: 'number',
  password: 'password',
  radio: 'radio',
  range: 'range',
  reset: 'reset',
  search: 'search',
  submit: 'submit',
  tel: 'tel',
  text: 'text',
  url: 'url',
};

storiesOf('Input', module)
  .addDecorator(withKnobs)
  .add('Input Demo', () => (
    <Input
      name="demo-input"
      disabled={boolean('Disable', false)}
      label={text('Label', 'Name')}
      errorDescription={text('Error Description', '')}
      placeholder={text('Placeholder', 'Placeholder')}
      readOnly={boolean('Read only', false)}
      requiredy={boolean('Required', false)}
      // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
      // @ts-ignore *
      type={select('Type', options, 'text')}
      maxLength={number('Max Length', 15)}
      onChange={action('onChange')}
    />
  ))
  .add('Input with max length', () => (
    <div>
      <Input
        name="random-input"
        label="Random"
        placeholder="Start typing"
        maxLength={5}
      />
      You can only enter a max of 5 words
    </div>
  ))
  .add('Input with error', () => (
    <Input
      name="error-input"
      label={'Username'}
      errorDescription={'Invalid input'}
    />
  ))
  .add('Readonly Input', () => (
    <Input
      name="readonly-input"
      label={text('Label', 'Name')}
      readOnly={true}
    />
  ));
