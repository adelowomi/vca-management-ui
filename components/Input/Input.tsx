import React from 'react';
import styled from 'styled-components';

import mergeClassNames from '../../helpers/mergeClassNames';
import theme from '../../styles/theme';

export const FormInputContainer = styled.div`
  display: flex;
  flex-direction: column;
  width: 'fit-content';
  padding: 8px 0px;
`;

export const Label = styled.label`
  margin-bottom: 8px;
  font-family: 'Avenir Next';
  font-weight: 500;
  color: ${theme.colors.greyishBrown};
`;

export type InputTypes =
  | 'button'
  | 'checkbox'
  | 'color'
  | 'email'
  | 'file'
  | 'hidden'
  | 'image'
  | 'number'
  | 'password'
  | 'radio'
  | 'range'
  | 'reset'
  | 'search'
  | 'submit'
  | 'tel'
  | 'text'
  | 'url';

export interface InputProp {
  /**
   * Add some additional styles
   */
  className?: string;
  /**
   * Boolean of whether to make the input disabled
   */
  disabled?: boolean;
  /**
   * Displays error text if available
   */
  errorDescription?: string;
  /**
   * The short hint displayed in the input before the user enters a value.
   */
  placeholder?: string;
  /**
   * It prevents the user from changing the value of the field.
   */
  readOnly?: boolean;
  /**
   * it specifies that an input field must be filled out
   */
  required?: boolean;
  /**
   * inputType
   * The HTML type of the input field
   * text, password, email, phone, etc.
   */
  type?: InputTypes;
  inputRef?: React.RefObject<HTMLInputElement>;
  /**
   * Determines label of input field
   */
  label?: string;
  /**
   * Defines the maximum number of characters the user can enter into the `input`.
   */
  maxLength?: number;
  /**
   *  name attribute specifies the name of the input element
   */
  name: string;
  onChange?: React.ChangeEventHandler<HTMLInputElement>;
}

const Input: React.FC<InputProp> = ({
  className,
  disabled,
  errorDescription,
  placeholder,
  readOnly,
  required,
  type,
  inputRef,
  label,
  maxLength,
  name,
  onChange,
  ...rest
}) => {
  const disabledClassName = disabled && 'app-disabled';
  const inputStyle = errorDescription
    ? 'shadow-sm focus:ring-red-500 focus:border-red-500 focus:border-2 block w-96 h-14 sm:text-sm border border-red-500 rounded-sm pl-4'
    : 'shadow-sm focus:ring-blue-500 focus:border-blue-500 focus:border-2 block w-96 h-14 sm:text-sm border border-gray-400 rounded-sm pl-4';
  return (
    <div className="flex flex-col px-0">
      <label htmlFor={name} className="mb-6 text-xl font-medium " id={label}>
        {label}
      </label>
      <input
        className={mergeClassNames(disabledClassName, className, inputStyle)}
        type={type}
        disabled={disabled}
        placeholder={placeholder}
        readOnly={readOnly}
        maxLength={maxLength}
        ref={inputRef}
        required={required}
        onChange={onChange}
        data-testid="input-test"
        // eslint-disable-next-line react/jsx-props-no-spreading
        {...rest}
      />
      {errorDescription && (
        <p className="pt-2 text-sm app-red">{errorDescription}</p>
      )}
    </div>
  );
};

const noOp = () => undefined;

Input.defaultProps = {
  className: '',
  disabled: false,
  label: '',
  errorDescription: '',
  placeholder: '',
  readOnly: false,
  required: false,
  type: 'text',
  maxLength: null,
  name: '',
  onChange: noOp,
};

export default Input;
