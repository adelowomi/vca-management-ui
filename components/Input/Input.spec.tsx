import '@testing-library/jest-dom';

import { render } from '@testing-library/react';
import React from 'react';

// Components
import Input from './Input';

describe('Input', () => {
  it('Renders input', () => {
    const { container } = render(<Input label="Default" name="Default" />);
    expect(container.firstChild).toMatchSnapshot();
  });
  it('Renders disabled input', () => {
    const { container } = render(
      <Input label="Default" name="Default" disabled={true} />
    );
    expect(container.firstChild).toMatchSnapshot();
  });
  it('Renders input with placeholder', () => {
    const { container } = render(
      <Input label="Default" name="Default" placeholder="Placeholder" />
    );
    expect(container.firstChild).toMatchSnapshot();
  });
  it('Renders readonly input', () => {
    const { container } = render(
      <Input label="Default" name="Default" readOnly={true} />
    );
    expect(container.firstChild).toMatchSnapshot();
  });
  it('Renders input with maxLength', () => {
    const { container } = render(
      <Input label="Default" name="Default" maxLength={6} />
    );
    expect(container.firstChild).toMatchSnapshot();
  });
  it('Renders error', () => {
    const { container } = render(
      <Input label="Default" name="Default" errorDescription="error" />
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
