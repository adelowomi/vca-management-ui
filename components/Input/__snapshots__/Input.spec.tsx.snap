// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Renders disabled input 1`] = `
<div
  class="flex flex-col w-full px-0"
>
  <label
    class="mb-2 text-sm font-medium app-greyish-brown"
    for="Default"
    id="Default"
  >
    Default
  </label>
  <input
    class="app-input app-disabled"
    data-testid="input-test"
    disabled=""
    placeholder=""
    type="text"
  />
  
</div>
`;

exports[`Input Renders error 1`] = `
<div
  class="flex flex-col w-full px-0"
>
  <label
    class="mb-2 text-sm font-medium app-greyish-brown"
    for="Default"
    id="Default"
  >
    Default
  </label>
  <input
    class="app-input"
    data-testid="input-test"
    placeholder=""
    type="text"
  />
  <p
    class="pt-2 text-sm app-red"
  >
    error
  </p>
</div>
`;

exports[`Input Renders input 1`] = `
<div
  class="flex flex-col w-full px-0"
>
  <label
    class="mb-2 text-sm font-medium app-greyish-brown"
    for="Default"
    id="Default"
  >
    Default
  </label>
  <input
    class="app-input"
    data-testid="input-test"
    placeholder=""
    type="text"
  />
  
</div>
`;

exports[`Input Renders input with maxLength 1`] = `
<div
  class="flex flex-col w-full px-0"
>
  <label
    class="mb-2 text-sm font-medium app-greyish-brown"
    for="Default"
    id="Default"
  >
    Default
  </label>
  <input
    class="app-input"
    data-testid="input-test"
    maxlength="6"
    placeholder=""
    type="text"
  />
  
</div>
`;

exports[`Input Renders input with placeholder 1`] = `
<div
  class="flex flex-col w-full px-0"
>
  <label
    class="mb-2 text-sm font-medium app-greyish-brown"
    for="Default"
    id="Default"
  >
    Default
  </label>
  <input
    class="app-input"
    data-testid="input-test"
    placeholder="Placeholder"
    type="text"
  />
  
</div>
`;

exports[`Input Renders readonly input 1`] = `
<div
  class="flex flex-col w-full px-0"
>
  <label
    class="mb-2 text-sm font-medium app-greyish-brown"
    for="Default"
    id="Default"
  >
    Default
  </label>
  <input
    class="app-input"
    data-testid="input-test"
    placeholder=""
    readonly=""
    type="text"
  />
  
</div>
`;
